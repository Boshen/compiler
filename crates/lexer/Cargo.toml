[package]
name = "lexer"
version = "0.0.0"
edition = "2021"

[dependencies]
unicode-id = "0.3.0"

[dev-dependencies]
insta = "1.12.0"
unindent = "0.1.7"
