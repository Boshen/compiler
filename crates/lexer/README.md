# Lexer

## Goal

### Performance
Performance should be achieved by reducing the total amount of if checks in any code path.

## Fuzzing

```bash
cd crates/lexer && cargo +nightly fuzz run lexer -- -only_ascii=1
```
