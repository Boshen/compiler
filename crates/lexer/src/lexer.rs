//! Lexer

use crate::kind::Kind;
use crate::token::Token;

pub struct Lexer<'a> {
    /// The input string
    bytes: &'a [u8],

    /// The current index position
    cur: usize,

    /// Are we at the EOF?
    eof: bool,
}

impl<'a> Lexer<'a> {
    #[must_use]
    pub const fn new(input: &'a str) -> Self {
        Self {
            bytes: input.as_bytes(),
            cur: 0,
            eof: false,
        }
    }
}

impl Iterator for Lexer<'_> {
    type Item = Token;

    fn next(&mut self) -> Option<Self::Item> {
        if self.cur >= self.bytes.len() {
            if self.eof {
                return None;
            }
            self.eof = true;
            return Some(Token::new(Kind::EOF, self.cur, 1));
        }

        let token = self
            .read_whitespace()
            .or_else(|| self.read_identifier())
            .or_else(|| Some(Token::new(Kind::Unknown, self.cur, 1)));

        self.cur += token.as_ref().unwrap().len();

        token
    }
}

impl Lexer<'_> {
    fn read_whitespace(&mut self) -> Option<Token> {
        let mut len = 0;

        for chr in &self.bytes[self.cur..] {
            if !chr.is_ascii_whitespace() {
                break;
            }
            len += 1;
        }

        if len == 0 {
            return None;
        }

        Some(Token::new(Kind::WhiteSpace, self.cur, len))
    }

    fn read_identifier(&mut self) -> Option<Token> {
        let mut len = 0;

        for chr in &self.bytes[self.cur..] {
            if !chr.is_ascii_alphabetic() {
                break;
            }
            len += 1;
        }

        if len == 0 {
            return None;
        }

        let kind = match &self.bytes[self.cur..self.cur + len] {
            b"null" => Kind::Null,
            b"undefined" => Kind::Undefined,
            b"true" => Kind::True,
            b"false" => Kind::False,
            _ => Kind::Ident,
        };

        Some(Token::new(kind, self.cur, len))
    }
}
