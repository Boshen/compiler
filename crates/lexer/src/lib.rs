#![warn(clippy::all, clippy::pedantic, clippy::nursery)]

mod constants;
mod kind;
mod lexer;
mod state;
mod token;

pub use crate::kind::*;
pub use crate::lexer::*;
pub use crate::token::*;
