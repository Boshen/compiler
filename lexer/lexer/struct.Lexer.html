<!DOCTYPE html>
<html lang="en">
<head>
    <meta http-equiv="refresh" content="0;URL=../../lexer/struct.Lexer.html">
    <title>Redirection</title>
</head>
<body>
    <p>Redirecting to <a href="../../lexer/struct.Lexer.html">../../lexer/struct.Lexer.html</a>...</p>
    <script>location.replace("../../lexer/struct.Lexer.html" + location.search + location.hash);</script>
</body>
</html>