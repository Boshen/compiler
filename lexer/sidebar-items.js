initSidebarItems({"enum":[["Kind",""],["Number",""]],"struct":[["Lexer",""],["Token",""]]});