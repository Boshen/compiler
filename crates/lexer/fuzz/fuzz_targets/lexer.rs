#![no_main]

use lexer::Lexer;
use libfuzzer_sys::fuzz_target;

fuzz_target!(|data: &[u8]| {
    if let Ok(s) = std::str::from_utf8(data) {
        Lexer::new(s).into_iter().count();
    }
});
