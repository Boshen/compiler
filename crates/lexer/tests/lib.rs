#![warn(clippy::all, clippy::pedantic, clippy::nursery)]
#[cfg(test)]
use lexer::Lexer;
use unindent::unindent;

fn test_snapshot(name: &str, input: &str) {
    let input = unindent(input);
    let tokens = Lexer::new(input.trim())
        .into_iter()
        .filter(|t| !t.kind().is_whitespace())
        .collect::<Vec<_>>();
    let snapshot = format!("# Input\n{}\n---\n# Output\n{:#?}", input, tokens);
    insta::with_settings!({
        prepend_module_to_snapshot => false,
    }, {
        insta::assert_snapshot!(name, snapshot, name);
    });
}

#[test]
fn comment() {
    let input = "
        // comment
        /* multiline comment */
    ";
    test_snapshot("comment", input);
}

#[test]
fn keyword() {
    let input = "
        await break case catch class const continue debugger default delete do else enum export extends
        false finally for function if import in instanceof new null return super switch this throw true try typeof var void while with yield
        undefined
    ";
    test_snapshot("keyword", input);
}

#[test]
fn identifier() {
    let input = r#"
        $ _ $a _a abc_$
        \u03bc
        \u{61}
    "#;
    test_snapshot("identifier", input);
}

#[test]
fn punctuator() {
    let input = "
        { ( ) [ ] . ... ; , < > <= >= == != === !== + - * % ** ++ -- << >> >>> & | ^ ! ~ && || ?? ? : = +=
        -= *= %= **= <<= >>= >>>= &= |= ^= &&= ||= ??= =>
        ?. / /= }
        #
    ";
    test_snapshot("punctuator", input);
}

#[test]
fn numeric_literal() {
    let input = "
        0 0n
        0b1 0B12
        0o1 0O12
        0x1 0X12
        0123 0789
        0. 0.0 0.123
        123n
        1.0 1.1
    ";
    test_snapshot("numeric_literal", input);
}

#[test]
fn string_literal() {
    let input = r#"
        "12345" '12345'
        "" ''
        '\\\n\r\t\b\v\f\'\"'
        '\\\n\r\t\b\v\f\'\"'
        '\u1234' '\x12'
        'foo \
        '
        "\d"
    "#;
    test_snapshot("string_literal", input);
}

#[test]
fn regex() {
    let input = r#"
        /aa/;
        /[0-9A-Za-z_\$(|)\[\]\/\\^]/;
        /[//]/;
        /[/]/;
    "#;
    test_snapshot("regex", input);
}

#[test]
fn template_literal() {
    let input = r#"
        ``
        `123`
        `\`\r`
    "#;
    test_snapshot("template_literal", input);
}
