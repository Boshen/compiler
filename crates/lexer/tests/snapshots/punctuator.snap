---
source: crates/lexer/tests/lib.rs
assertion_line: 13
expression: punctuator

---
# Input
{ ( ) [ ] . ... ; , < > <= >= == != === !== + - * % ** ++ -- << >> >>> & | ^ ! ~ && || ?? ? : = +=
-= *= %= **= <<= >>= >>>= &= |= ^= &&= ||= ??= =>
?. / /= }
#

---
# Output
[
    Token {
        kind: LCurly,
        offset: 0,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 1,
        len: 1,
    },
    Token {
        kind: LParen,
        offset: 2,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 3,
        len: 1,
    },
    Token {
        kind: RParen,
        offset: 4,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 5,
        len: 1,
    },
    Token {
        kind: LBrack,
        offset: 6,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 7,
        len: 1,
    },
    Token {
        kind: RBrack,
        offset: 8,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 9,
        len: 1,
    },
    Token {
        kind: Dot,
        offset: 10,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 11,
        len: 1,
    },
    Token {
        kind: Dot3,
        offset: 12,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 15,
        len: 1,
    },
    Token {
        kind: Semicolon,
        offset: 16,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 17,
        len: 1,
    },
    Token {
        kind: Comma,
        offset: 18,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 19,
        len: 1,
    },
    Token {
        kind: LAngle,
        offset: 20,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 21,
        len: 1,
    },
    Token {
        kind: RAngle,
        offset: 22,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 23,
        len: 1,
    },
    Token {
        kind: LtEq,
        offset: 24,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 26,
        len: 1,
    },
    Token {
        kind: GtEq,
        offset: 27,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 29,
        len: 1,
    },
    Token {
        kind: Eq2,
        offset: 30,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 32,
        len: 1,
    },
    Token {
        kind: Neq,
        offset: 33,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 35,
        len: 1,
    },
    Token {
        kind: Eq3,
        offset: 36,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 39,
        len: 1,
    },
    Token {
        kind: Neq2,
        offset: 40,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 43,
        len: 1,
    },
    Token {
        kind: Plus,
        offset: 44,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 45,
        len: 1,
    },
    Token {
        kind: Minus,
        offset: 46,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 47,
        len: 1,
    },
    Token {
        kind: Star,
        offset: 48,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 49,
        len: 1,
    },
    Token {
        kind: Percent,
        offset: 50,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 51,
        len: 1,
    },
    Token {
        kind: Star2,
        offset: 52,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 54,
        len: 1,
    },
    Token {
        kind: Plus2,
        offset: 55,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 57,
        len: 1,
    },
    Token {
        kind: Minus2,
        offset: 58,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 60,
        len: 1,
    },
    Token {
        kind: ShiftLeft,
        offset: 61,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 63,
        len: 1,
    },
    Token {
        kind: ShiftRight,
        offset: 64,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 66,
        len: 1,
    },
    Token {
        kind: ShiftRight3,
        offset: 67,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 70,
        len: 1,
    },
    Token {
        kind: Amp,
        offset: 71,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 72,
        len: 1,
    },
    Token {
        kind: Pipe,
        offset: 73,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 74,
        len: 1,
    },
    Token {
        kind: Caret,
        offset: 75,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 76,
        len: 1,
    },
    Token {
        kind: Bang,
        offset: 77,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 78,
        len: 1,
    },
    Token {
        kind: Tilde,
        offset: 79,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 80,
        len: 1,
    },
    Token {
        kind: Amp2,
        offset: 81,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 83,
        len: 1,
    },
    Token {
        kind: Pipe2,
        offset: 84,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 86,
        len: 1,
    },
    Token {
        kind: Question2,
        offset: 87,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 89,
        len: 1,
    },
    Token {
        kind: Question,
        offset: 90,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 91,
        len: 1,
    },
    Token {
        kind: Colon,
        offset: 92,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 93,
        len: 1,
    },
    Token {
        kind: Eq,
        offset: 94,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 95,
        len: 1,
    },
    Token {
        kind: PlusEq,
        offset: 96,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 98,
        len: 1,
    },
    Token {
        kind: MinusEq,
        offset: 99,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 101,
        len: 1,
    },
    Token {
        kind: StarEq,
        offset: 102,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 104,
        len: 1,
    },
    Token {
        kind: PercentEq,
        offset: 105,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 107,
        len: 1,
    },
    Token {
        kind: Star2Eq,
        offset: 108,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 111,
        len: 1,
    },
    Token {
        kind: ShiftLeftEq,
        offset: 112,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 115,
        len: 1,
    },
    Token {
        kind: ShiftRightEq,
        offset: 116,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 119,
        len: 1,
    },
    Token {
        kind: ShiftRight3Eq,
        offset: 120,
        len: 4,
    },
    Token {
        kind: WhiteSpace,
        offset: 124,
        len: 1,
    },
    Token {
        kind: AmpEq,
        offset: 125,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 127,
        len: 1,
    },
    Token {
        kind: PipeEq,
        offset: 128,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 130,
        len: 1,
    },
    Token {
        kind: CaretEq,
        offset: 131,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 133,
        len: 1,
    },
    Token {
        kind: Amp2Eq,
        offset: 134,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 137,
        len: 1,
    },
    Token {
        kind: Pipe2Eq,
        offset: 138,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 141,
        len: 1,
    },
    Token {
        kind: Question2Eq,
        offset: 142,
        len: 3,
    },
    Token {
        kind: WhiteSpace,
        offset: 145,
        len: 1,
    },
    Token {
        kind: FatArrow,
        offset: 146,
        len: 2,
    },
    Token {
        kind: WhiteSpace,
        offset: 148,
        len: 1,
    },
    Token {
        kind: Question,
        offset: 149,
        len: 1,
    },
    Token {
        kind: Dot,
        offset: 150,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 151,
        len: 1,
    },
    Token {
        kind: Regex,
        offset: 152,
        len: 3,
    },
    Token {
        kind: Eq,
        offset: 155,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 156,
        len: 1,
    },
    Token {
        kind: RCurly,
        offset: 157,
        len: 1,
    },
    Token {
        kind: WhiteSpace,
        offset: 158,
        len: 1,
    },
    Token {
        kind: Hash,
        offset: 159,
        len: 1,
    },
    Token {
        kind: EOF,
        offset: 160,
        len: 1,
    },
]
